### PR TITLE
Protect against missing track info in MiniAOD

### PR DESCRIPTION
**Pull request by ahinzmann**
When pileup jet ID is reran on top of an existing MiniAOD, it crashes due to missing tracking information.
The fix sets the corresponding variables to default values if no information is available.
The affected variables are not used in the default MVA training of pileup jet ID, thus have no effect on the observable used in analyses.
Backport to 93 for analyses: #20740

### CODE REVIEW

**Comment by slava77** [comment]
passing code-checks is required.
Please apply the patch

File: RecoJets/JetProducers/src/PileupJetIdAlgo.cc [hunk 1]
```diff
@@ -393,10 +393,19 @@ PileupJetIdentifier PileupJetIdAlgo::computeIdVariables(const reco::Jet * jet, f
 			    //To handle the electron case
 			    if(lPF!=nullptr) {
 			      pfTrk=(lPF->trackRef().get()==nullptr)?lPF->gsfTrackRef().get():lPF->trackRef().get();
+			      internalId_.d0_ = std::abs(pfTrk->dxy(vtx->position()));
+			      internalId_.dZ_ = std::abs(pfTrk->dz(vtx->position()));
+			    }
+			    if(lPack!=nullptr) {
```
Comment (perrotta) [inline RecoJets/JetProducers/src/PileupJetIdAlgo.cc]: you already have an isPacked bool variable that you can use without needing to check the existence of the pointer every time

File: RecoJets/JetProducers/src/PileupJetIdAlgo.cc [hunk 2]
```diff
@@ -393,10 +393,19 @@ PileupJetIdentifier PileupJetIdAlgo::computeIdVariables(const reco::Jet * jet, f
 			    //To handle the electron case
 			    if(lPF!=nullptr) {
```
Comment (perrotta) [inline RecoJets/JetProducers/src/PileupJetIdAlgo.cc]: Shouldn't this be considered with an "else if"  after the block started at L399?
With the present logic, if (isPacked) the d0 and dz saved here will be rewritten anyhow

**Comment by perrotta** [comment]
@ahinzmann : you still need to apply the code-checks, as explained in https://github.com/cms-sw/cmssw/pull/20741#issuecomment-335155731. Without them the tests cannot be launched in github.

**Comment by perrotta** [comment]
please test

File: RecoJets/JetProducers/src/PileupJetIdAlgo.cc [hunk 3]
```diff
@@ -391,12 +391,21 @@ PileupJetIdentifier PileupJetIdAlgo::computeIdVariables(const reco::Jet * jet, f
 			  }
 			  if(pfTrk==nullptr) { //protection against empty pointers for the miniAOD case
 			    //To handle the electron case
-			    if(lPF!=nullptr) {
+			    if(isPacked) {
```
Comment (gpetruc) [inline RecoJets/JetProducers/src/PileupJetIdAlgo.cc]: given that #20491 is merged, you could use the `PackedCandidate::dz(point)` and `PackedCandidate::dxy(point)` methods, which are available for all particles (`hasTrackDetails` is only needed to get the covariance, hitpattern or other more detailed information)

Comment (ahinzmann) [reply RecoJets/JetProducers/src/PileupJetIdAlgo.cc]: ok, I will change this. Was #20491 also backported to 93?

**Comment by perrotta** [comment]
@ahinzmann : no, there is so far no backport of #20491 to 93X

By the way: since if I am not wrong the issue was already present in 92X, do you also plan/need any backport in that release serie?

**Comment by ahinzmann** [comment]
ok, let's also backport to 92 once the master is merged

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
Matrix error failure is unrelated, likely due to the merging of #20693

**Comment by perrotta** [comment]
please test
(the latest IB, CMSSW_9_4_X_2017-10-11-2300, should finally contain the fix for the issue in the RelVal wf 1000)

**Comment by perrotta** [comment]
+1
- Protection added (implicitly, taking profit of  #20491 merged in 94X) for cases in which the tracks are not available
- It only matters for pileup jet ID rerun on top of an existing MiniAOD: no changes expected for standard workflows
- No effect on jenkins outputs, as expected

**Comment by davidlange6** [comment]
+1

**Pull request merged**